atmosphere/plugins/machine_validation.py

```diff
@@ -0,0 +1,140 @@
+from threepio import logger
+
+
+class MachineValidationPlugin(object):
+    def __init__(self, account_driver):
+        self.account_driver = account_driver
+
+    def machine_is_valid(self, cloud_machine):
+        raise NotImplementedError(
+            "Validation plugins must implement a machine_is_valid function "
+            "that takes arguments: 'accounts', 'cloud_machine'")
+
+    def _sanity_check_machine(self, cloud_machine):
+        """
+        The following sanity checks should be done on any image, before adding to
+        the catalog:
+
+        - Fail if image is None
+        - Fail if image is a 'snapshot'
+        - Fail if image is a kernel or ramdisk
+        """
+        # Skip the machine if the owner was not the chromogenic image creator
+        # (tenant name must match admin tenant name)
+        # If that behavior changes, this snippet should be updated/removed.
+        if not cloud_machine:
+            return False
+        elif self._is_kernel_or_ramdisk(cloud_machine):
+            logger.info("Skipping cloud machine %s - kernel/ramdisk found" % cloud_machine)
+            return False
+        elif self._is_snapshot(cloud_machine):
+            logger.info("Skipping cloud machine %s - snapshot found" % cloud_machine)
+            return False
+        return True
+
+    def _contains_metadata(self, cloud_machine, metadata_key):
+        return True if hasattr(cloud_machine, metadata_key) else False
+
+    def _machine_authored_by_atmosphere(self, cloud_machine):
+        project_id = cloud_machine.get('owner')
+        owner_project = self.account_driver.get_project_by_id(project_id)
+        if not owner_project:
+            owner = cloud_machine.get('application_owner')
+            owner_project = self.account_driver.get_project(owner)
+        # Assumption: the atmosphere imaging author == the project_name set for the account_driver.
+        atmo_author_project_name = self.account_driver.project_name
+        if not owner_project:
+            logger.info(
+                "cloud machine %s - authored by project_id %s, not the Atmosphere author: %s",
+                cloud_machine.id, project_id, atmo_author_project_name)
+            return False
+        elif owner_project.name != atmo_author_project_name:
+            logger.info(
+                "cloud machine %s - authored by Tenant %s, not the Atmosphere author: %s",
+                cloud_machine.id, owner_project.name, atmo_author_project_name)
+            return False
+        return True
+
+    def _is_kernel_or_ramdisk(self, cloud_machine):
+        cloud_machine_name = cloud_machine.name if cloud_machine.name else ""
+        if any(cloud_machine_name.startswith(prefix) for prefix in ['eri-', 'eki-', 'ari-', 'aki-']):
```

Comment:
-1 this sets a trap for users. My friend at the Edmonton Research Institute is about to create his "eri-genomics" image, and then become sad when it doesn't show up in Atmosphere.

Reply:
Good point, i'll drop the name restrictions.